### ISSUE TITLE
Data loading (in utils/datasets.py) is quite ugly, although highly functional. 

### ISSUE THREAD

**Issue by learning-luke** [BayesWatch/pytorch-experiments-template#4]
This could do with an overhaul and extension, particularly to new datasets like imagenet, etc. 
The way that I check for which dataset is to be used is also ugly.

**Comment by jack-willturner** [comment]
OK my main criticism of the current data loading situation is that there is too much control flow. 

E.g. if I want to make a slight modification to MNIST - there are ~4 different if statements I might have to edit. I'd prefer all of the data loading for MNIST to be in one function. 

So I propose something like:

```python
def get_dataset(args):
    if args.dataset == 'mnist':
        return get_mnist_loaders(args)
    elif args.dataset == 'cifar10':
        return get_cifar10_loaders(args)
    elif args.dataset == 'cifar100':
        return get_cifar100_loaders(args)
    elif args.dataset == 'imagenet':
        return get_imagenet_loaders(args)
```

Then to add a new dataset, I just have to add one function - I don't have to go into the loader and figure out which bits of control flow apply to me or worry about messing with the other datasets.

The downside to this is lots of replication. Each of these functions will look like:

```python
def get_cifar10_loaders(args):

    normalize = transforms.Normalize(mean=[0.4914, 0.4822, 0.4465],
                                     std=[0.2023, 0.1994, 0.2010])

    transform_train = transforms.Compose([
        transforms.RandomCrop(32, padding=4),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        normalize,
    ])

    transform_validate = transforms.Compose([
        transforms.ToTensor(),
        normalize,
    ])

    train_set = torchvision.datasets.CIFAR10(...)
    val_set = torchvision.datasets.CIFAR10(...)

    train_loader = torch.utils.data.DataLoader(...)
    val_loader = torch.utils.data.DataLoader(...)
    return train_loader, val_loader
```

Then if I wanted to add e.g. a new augmentation and test on all the datasets, I would have to add the new augmentation to every one of the functions. 

Thoughts?

**Comment by AntreasAntoniou** [comment]
Looks good to me. Just one point. Let's try to avoid args as the input to anything other than the main train.py script. It makes the code much more readable. For example here, use dataset_name as the input to get_dataset.

**Comment by AntreasAntoniou** [comment]
I think here replication might be a good thing. One can abstract this to a more flexible variant quite easily if necessary. But this is a good level of abstraction I think.

**Comment by jack-willturner** [comment]
OK a first attempt at this: 31f2db2. It's a proof-of-concept and is lacking a lot of the functionality the original data loader has.

The ImageNet data loader doesn't work out of the box because we also need to make architectural changes (changing the size of the pooling layer and maybe some strides) when moving from ~CIFAR-size to ImageNet-size. 

I'm not sure what the best way to handle this is going to be.

**Comment by jack-willturner** [comment]
Another thought: each of the data loaders implicitly implement an abstract base class (i.e. they all have a `transform_train`, `transform_validate`, `im_size`, `load_dataset`). 

I could add an ABC to enforce the implementation of these four things whenever we add a new dataset but I worry this might be over-engineering.

**Comment by learning-luke** [comment]
I don't think we need that last, Jack. But this looks so much better now.

**Comment by learning-luke** [comment]
One problem now is that the removal of some functionality. For instance, the CINIC10 downloader.

**Comment by jack-willturner** [comment]
The functionality I have cut:

1. CINIC downloading 
2. Various augmentation flags (colour jitter, shears, cutout etc.)
3. Fashion MNIST (easy to add back in but not sure how often it's used?)
4. Custom means/stds via argparse 
5. The validation set maker 
6. The subset sequential sampler (though I'm not sure exactly why this is needed?)

I think that's everything.

I'm happy to add all of this back in if you miss it (and that would probably be a good test of how extensible this is).

Have you considered opening a PR to add CINIC to torchvision datasets (so we could just have a "download=True" flag?)

**Comment by learning-luke** [comment]
I think having access to augmentations would be really nice. BUT, I think that we should build these as there own "groups" so to speak. For instance, there are a set of augmentations that are used in Contrastive Learning (like SIMClr) and then there are also Fast and Auto Augment. Being able to simple switch those transforms off and on (abstracted to another class) would be amazing. 

Also, I think that any work on modern unsupervised learning would typically require contrastive approaches. And, that means that we need both the original image and the augmented image. This is annoying to do with the different ways that CIFAR and MNIST and the Dataset default work with data, but being able to do this would be amazing. My way is messy. Any ideas?

**Comment by jack-willturner** [comment]
It should be easy enough to write a generic contrastive loader for any dataset:

```python
class ContrastiveLoader():
    def __init__(self, loader=CIFAR10Loader):
        base_loader = loader()
        contrastive_loader = loader()

        ## set contrastive loading scheme 
        contrastive_loader.transform_train = ...
    
    def get_data(self, data_loc):
        base_train_set, base_val_set = base_loader.get_data(data_loc)
        contrastive_train_set, contrastive_val_set = ontrastive_loader.get_data(data_loc)

        return base_train_set, base_val_set, contrastive_train_set, contrastive_val_set   
```

As long as you can seed the two data loaders such that they shuffle in exactly the same order.

Is this normally how you do contrastive loaders? I've never used them before - there's obviously a far more memory efficient way to do this.

**Comment by jack-willturner** [comment]
As a proof-of-concept I implemented how I think contrastive data loading should work here: 4076591. It's really just one line of code 😄 

There's a Jupyter notebook so you can see it works too, which I'll delete before we do any merging. 

Let me know your thoughts.

**Comment by AntreasAntoniou** [comment]
I added some comments to both your proposals. Overall they look awesome. Let me know your thoughts in my comments.

**Comment by jack-willturner** [comment]
OK I think I've addressed your comments! Namely:
- made sure functions were called with named arguments
- switched `workers` to `num_dataloading_workers` 

I've also:
- reintroduced data downloading
- moved SimCLR into `augmentations.py` so that the various extra aug schemes can be grouped together

If you are both happy with the current status, I'll delete the old `datasets.py` and open a PR.

**Comment by AntreasAntoniou** [comment]
I don't like how num_classes is currently set. It's too hacky/manual. Ideally the data provider should be used as the source of that information. Do you guys mind if we return:

```
trainloader, train_dataset, testloader, test_dataset, in_shape = load_dataset(args)
```

instead of 

```
trainloader, testloader, in_shape = load_dataset(args)
```

**Comment by learning-luke** [comment]
@jack-willturner , have you tested that CINIC is working with your reimplementation?

**Comment by jack-willturner** [comment]
@learning-luke just checked: I had an ssl issue which I've temporarily fixed. There's probably a better solution but at least for the moment it works.

**Comment by jack-willturner** [comment]
> I don't like how num_classes is currently set. It's too hacky/manual. Ideally the data provider should be used as the source of that information. Do you guys mind if we return:
> 
> ```
> trainloader, train_dataset, testloader, test_dataset, in_shape = load_dataset(args)
> ```
> 
> instead of
> 
> ```
> trainloader, testloader, in_shape = load_dataset(args)
> ```

This is fixed now.
